mirrord/layer/src/socket/ops.rs

```diff
@@ -191,15 +177,68 @@ pub(super) fn bind(
         Err(HookError::AddressAlreadyBound(requested_address))?;
     }
 
-    let bind_result = unsafe { FN_BIND(sockfd, unbound_address.as_ptr(), unbound_address.len()) };
+    // Try to bind a port from listen ports, if no configuration
+    // try to bind the requested port, if not available get a random port
+    // if there's configuration and binding fails with the requested port
+    // we return address not available and not fallback to a random port.
+    let listen_port = LISTEN_PORTS
+        .get()
+        .expect("`LISTEN_PORTS` not set. Please report a bug")
+        .get_by_left(&requested_address.port())
+        .cloned()
+        .unwrap_or(requested_address.port());
+
+    let unbound_address = match socket.domain {
+        libc::AF_INET => Ok(SockAddr::from(SocketAddr::new(
+            IpAddr::V4(Ipv4Addr::LOCALHOST),
+            listen_port,
+        ))),
+        libc::AF_INET6 => Ok(SockAddr::from(SocketAddr::new(
+            IpAddr::V6(Ipv6Addr::UNSPECIFIED),
+            listen_port,
+        ))),
+        invalid => Err(Bypass::Domain(invalid)),
+    }?;
+
+    trace!("bind -> unbound_address {:#?}", unbound_address);
+
+    let mut bind_result =
+        unsafe { FN_BIND(sockfd, unbound_address.as_ptr(), unbound_address.len()) };
     if bind_result != 0 {
-        error!(
-            "listen -> Failed `bind` sockfd {:#?} to address {:#?} with errno {:#?}!",
-            sockfd,
-            unbound_address,
-            errno::errno()
-        );
-        return Err(io::Error::last_os_error())?;
+        // If we didn't have `listen_ports` used here, just assign a random address.
```

Comment:
You could extract this whole thing here into a shared `call_raw_bind(listen_port)`, and have a more _rusty_ block of code.

Like the code before would be something like:

```rust
let bound = do_the_bind(listen_port).or_else(|e| { 
    warn!("..."); // as you don't like `inspect_err` much
    do_the_bind(0)
})?;
```